### DISCUSSION TITLE
Use `fastapi` CLI but infer the port to listen to at runtime

### DISCUSSION THREAD

**Discussion by galah92** [fastapi/fastapi#12104] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
`uv run fastapi run main.py`
```

### Description

When running the `fastapi` CLI the default port is 8000, which can be overwritten with the `--port` flag. I'm deploying my service to Cloud Run, which specify the port to listen to at runtime. How do I support that? I'd like to keep using `fastapi` CLI since it's the best practice right now.

### Operating System

Linux

### FastAPI Version

0.112.2

### Pydantic Version

2.8.2

### Python Version

Python 3.12.5

**Comment by meower1** [comment]
I'm not sure if i understood correctly, i assume you mean that whenever you execute your python file using lets say 
`python3 main.py` you'd want your fastapi application to be ran. if thats the case you can go ahead and implement something like this

```py


from fastapi import FastAPI

app = FastAPI()

@app.get("/")
async def router_name() -> str:
    return "hello world"

if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, port=specified_port)
```

also the fastapi cli is built on top of uvicorn so its not necessarily best practice in my opinion.

**Reply by galah92** [reply]
I want to use `fastapi` CLI, under the assumption it'll get updates according to industry standard. And I want it to pick the right port from environment variable.

**Comment by meower1** [comment]
well in that case you can just modify the code to run fastapi cli but now you got more dependencies to import

```py
from fastapi import FastAPI

app = FastAPI()


@app.get("/")
async def router_name() -> str:
    return f"hello world"


if __name__ == "__main__":

    from pathlib import Path
    import os

    current_dir = Path(__file__).parent.resolve()
    app_file = current_dir / "main.py"

    os.system(f"fastapi run {app_file} --port 8080")
```

**Reply by galah92** [reply]
Spawning a new process looks suboptimal. My request is to have a built-in support for this in `fastapi` CLI.

**Comment by slobodaapl** [comment]
I agree, this should be a basic functionality.. Take the port depending on what's defined at runtime:

1. If ENV var not available and flag not passed to cli command, use default
2. If flag passed, use that
3. If flag not passed, but ENV var available, use that

**Comment by ZlatyChlapec** [comment]
Seems like there is already [PR](https://github.com/fastapi/fastapi-cli/pull/18) for that, it only needs merging.

**Comment by alv2017** [comment]
FastAPI CLI supports --host and --port options. Please see the example below:

```
fastapi dev --host 127.0.0.1 --port 5555 main.py
```

In case you want to know all the supported options for `fastapi dev`:
```
fastapi dev --help
```

In case you want to know all the supported options for `fastapi run`:
```
fastapi run --help
```

**Comment by negentropicdev** [comment]
I would also like this ability at run-time. I have a library/framework that reads a configuration file shared amongst multiple applications in multiple languages which specifies things like http ports (if configured) to run with. Specifying via the CLI doesn't fit my use case. I'm working on using FastAPI within this framework, not as a top level component.
